elastic/apm-aws-lambda
# Fetch Lambda Platform Metrics

### Motivation / Summary
Following issue #108, the goal of this pull request is to send the following metrics to Elasticsearch through the AWS Lambda APM extension : 
- Init Duration
- Total Duration
- Billed Duration
- Memory Used
- Memory Sized

This information is typically provided by the [Lambda Platform Logs API](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-logs-api.html) as follows : 
```
REPORT RequestId: cc18960e-80e4-4160-bfdc-a5dae659993b	Duration: 162.81 ms	Billed Duration: 163 ms	Memory Size: 128 MB	Max Memory Used: 79 MB	Init Duration: 646.33 ms
```
Current demo dashboard : 
![image](https://user-images.githubusercontent.com/48380853/154601506-21838952-f001-408a-8bb6-99bd415f0fe9.png)

Memory usage is also displayed in the APM UI : 
<img width="1724" alt="image" src="https://user-images.githubusercontent.com/48380853/154602295-78268622-77bd-4576-9742-cd363e7298c4.png">

### Files / Added Elements
- A new file, `process_metrics.go` has been added to the extension (as well as the corresponding test file). This file contains the `structs` required to implements the Metrics model. Most of the code comes from the Go Agent codebase, with some simplifications (most notably at the Metrics aggregation level).
- In the `logs_api` folder, the `client` and the `http_listener` have been modified to take events of type `Report` into account.
- The main function has been modified to call the `process_metrics` functions upon dedication of a `Report` event.

### Limitations / Elements to discuss
- The Platform Logs report is only made available *after* the function is over. As such, metrics are only sent if another invocation is triggered in a timespan short enough for the function to leave its warm state. As per definition, all workarounds would extend the Lambda runtime and thus incur additional billing for users.
- Metadata retrieval is currently very limited, and done minimally through a mix of hard-coded values fitting the test example and environment variables. There are several ways to populate the metadata - create a dedicated endpoint on the extension for the agents to send the information, or decompress and analyse the first received transaction.
- The newly created fields have to be added manually to the APM index pattern for reproduction.

### Checklist

- [x] Provide a first proof of concept
- [x] Devise and implement a strategy for metadata retrieval
- [ ] Confirm the names of the newly added fields
- [ ] Refine proof of concept and dashboard
- [x] Settle on a final list of metrics to collect and expand scope
- [ ] Expose Lambda specific metrics via UI

### How to test
0. Make sure that your Go Runtime is at least 1.17
1. Clone/fork the current repository and switch to the following [branch](https://github.com/jlvoiseux/apm-aws-lambda?organization=jlvoiseux&organization=jlvoiseux)
2. Build a new version of the extension and save it as a layer
3. [Instrument](https://www.elastic.co/guide/en/apm/guide/current/aws-lambda-extension.html) the Lambda function of your choice
4. On your Elastic instance, add the following fields to the APM index pattern with the following configuration : 
- `faas.metrics.duration.init.ms`, Type `double`, Format `Duration` (Milliseconds)
- `faas.metrics.duration.measured.ms`, Type `double`, Format `Duration` (Milliseconds)
- `faas.metrics.duration.billed.ms`, Type `long`, Format `Duration` (Milliseconds)
- `faas.metrics.memory.maxUsed.bytes`, Type `long`, Format `Bytes` (b - Binary)
- `faas.metrics.memory.total.bytes`, Type `long`, Format `Bytes` (b - Binary)
5. Run the Lambda function several times as to trigger the fetch of several reports.
6. Observe the data in Elastic
